plugins/engines/ospray/ispc/render/PathTracingRenderer.ispc

```diff
@@ -0,0 +1,225 @@
+            break;
+        }
+        DifferentialGeometry dg;
+        postIntersect(self->abstract.super.model, dg, localray,
+                      DG_NG | DG_NS | DG_NORMALIZE | DG_FACEFORWARD |
+                          DG_MATERIALID | DG_COLOR);
+
+        uniform Material* material = dg.material;
+        uniform ExtendedOBJMaterial* objMaterial =
+            (uniform ExtendedOBJMaterial*)material;
+        if (!objMaterial)
+            Kd = make_vec3f(dg.color);
+        else
+            foreach_unique(mat in objMaterial) Kd =
+                mat->Kd * make_vec3f(dg.color);
+
+        // compute ray direction of uniformly random diffuse ray
+        const vec3f raydir =
+            getRandomVector(sample, dg.Ns, self->abstract.super.epsilon);
+
+        // origin of new ray in path is hitpoint of previous ray in path
+        const vec3f hitpoint = dg.P + self->abstract.super.epsilon * dg.Ns;
+        setRay(localray, hitpoint, raydir);
+        localray.t0 = self->abstract.super.epsilon;
+        localray.t = self->abstract.ambientOcclusionDistance;
+
+        if (self->abstract.shadows > 0.f)
+        {
+            // Direct lighting
+            for (uniform int i = 0;
+                 self->abstract.lights && i < self->abstract.numLights; ++i)
+            {
+                const uniform Light* uniform light = self->abstract.lights[i];
+                const vec2f s = make_vec2f(0.5f);
+                const varying Light_SampleRes lightSample =
+                    light->sample(light, dg, s);
+                const vec3f radiance = lightSample.weight;
+                if (reduce_max(radiance) > 0.f)
+                {
+                    const vec3f lightDirection = lightSample.dir;
+                    const vec3f sunSampleDirection =
+                        getConeSample(lightDirection, rng,
+                                      self->abstract.softShadows);
+                    const float sunLight =
+                        dot(dg.Ns,
+                            sunSampleDirection); // cosine weighted contribution
+
+                    if (sunLight > 0.f) // if surface is facing light
+                    {
+                        // create and trace a shadowray towards the sun
+                        Ray sunRay;
+                        setRay(sunRay, hitpoint, sunSampleDirection);
+                        if (isOccluded(self->abstract.super.model, sunRay))
+                            // take surface color at this bounce's hitpoint into
+                            // account
+                            Kd = Kd * (1.f - self->abstract.shadows);
+                    }
+                    else
+                        Kd = Kd * (1.f - self->abstract.shadows);
+                }
+            }
+        }
+        mask = mask * Kd;
+    } // end bounces
+
+    // CHECK THIS!!! WITH OR WITHOUT * MASK
+    // results are very different. Color bleeding is much stronger with mask
+    color = accucolor * mask;
```

Comment:
After some digging into path tracing theory, it would be more correct to not multiply the final color by the mask colour again, as it has already been taken into account. Otherwise this will result into too much colour bleeding.

Reply:
Ok. Fixed! :)